Maze.cpp

```diff
@@ -40,11 +44,196 @@ void Maze::setWall(CellCoordinate pos, Direction dir, bool wall) {
 }
 
 
-Path Maze::findPath(CellCoordinate start, CellCoordinate end)
+Path Maze::findPath(CellCoordinate start, CellCoordinate end, Direction facing )
+{
+	Node * startNode = getNode(start);
+
+	// The nodes that still need to be evaluated.
+	// Initially insert the start node.
+	set<Node *> openNodes;
+	openNodes.insert(getNode(start));
+	
+	// Reset any metadata from previous pathfinding.
+	resetNodePathData();
+	
+	// The start node is 0 distance away.
+	startNode->gScore = 0;
+	startNode->fScore = heuristic(start, end);
+	startNode->direction = facing;
+
+	// While there are node remaining to be evaluated.
+	while (!openNodes.empty())
+	{
+		// Get the open node with lowest score.
+		auto currentNodeItr = openNodes.begin();
+
+		// If the current node is the finish node.
+		if ( *currentNodeItr == getNode( end ) )
+		{
+			// We are done. Construct the path.
+			return constructPath(*currentNodeItr);
+		}
+		
+		// Set the current node as evaluated.
+		openNodes.erase(currentNodeItr);
+		(*currentNodeItr)->evaluated = true;
+
+		// For each node adjacent to the current node.
+		for (
+			Direction direction = Direction::N;
+			direction != Direction::NONE;
+#ifdef MAZE_DIAGONALS
+			direction = static_cast<Direction>(direction + 1)
+#else
+			direction = static_cast<Direction>(direction + 2)
+#endif // MAZE_DIAGONALS
+			)
+		{
+			Node * adjacentNode = getAdjacentNode(*currentNodeItr, direction);
+
+			// If the adjacent node has already been evaluated or does not exist.
+			if (adjacentNode->evaluated || !adjacentNode->exists)
```

Comment:
Then it isn't a problem for `->exists` here to always be true, even when there are walls around that corner?

Or, you said they should not exist; in which case, this condition will always be met, and the pathfinder will never use diagonals, right?